executor/src/witgen/machines/multiplicity_counter.rs

```diff
@@ -0,0 +1,73 @@
+use std::collections::BTreeMap;
+
+use powdr_ast::analyzed::PolyID;
+use powdr_number::{DegreeType, FieldElement};
+
+/// Utility for counting the multiplicities for lookups.
+pub struct MultiplicityCounter {
+    /// Maps an identity ID to the corresponding multiplicity column.
+    /// Note that multiple identity IDs can map to the same multiplicity column.
+    identity_id_to_multiplicity_column: BTreeMap<u64, PolyID>,
+
+    /// A map poly_id -> (index -> count).
+    /// Has a (possibly entry) entry for each value in `identity_id_to_multiplicity_column`.
+    counts: BTreeMap<PolyID, BTreeMap<usize, usize>>,
```

Comment:
I might be better to use `Vec` instead of a `BTreeMap` for the inner type. Did you test that?

Reply:
Hmm, a Vec that would be resized on demand (because we don't know the number of rows a priori)?

I have not tested this, but:
- The counts should be accessed at most once per machine call, so I wouldn't expect a `BTreeMap` access to be a bottleneck.
- I'd expect the vector to be sparse in practice. We typically use permutations to connect machines, so in practice I'd expect LogUp to be used only for fixed lookups (typically of size $2^{16}$ to $2^{18}$) and global range constraints. Only the former is handled in this PR (range constraints will need another pass over the data in an upcoming PR).